installer/digital_ocean_cluster.go

```diff
@@ -0,0 +1,462 @@
+		}
+	}
+	return nil, errors.New("Unable to locate create action ID")
+}
+
+func (c *DigitalOceanCluster) waitForActionComplete(actionID int) error {
+	fetchAction := func() (*godo.Action, error) {
+		action, _, err := c.client.Actions.Get(actionID)
+		if err != nil {
+			return nil, err
+		}
+		return action, nil
+	}
+	for {
+		action, err := fetchAction()
+		if err != nil {
+			return err
+		}
+		switch action.Status {
+		case "completed":
+			return nil
+		case "errored":
+			return errors.New("Droplet create failed")
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (c *DigitalOceanCluster) fetchInstanceIPs() error {
+	droplets, err := c.fetchDroplets()
+	if err != nil {
+		return err
+	}
+	instanceIPs := make([]string, 0, len(droplets))
+	for _, d := range droplets {
+		for _, n := range d.Networks.V4 {
+			instanceIPs = append(instanceIPs, n.IPAddress)
+		}
+	}
+	if int64(len(instanceIPs)) != c.base.NumInstances {
+		return fmt.Errorf("Expected %d instances, but found %d", c.base.NumInstances, len(instanceIPs))
+	}
+	c.base.InstanceIPs = instanceIPs
+	if err := c.base.saveInstanceIPs(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *DigitalOceanCluster) fetchDroplets() ([]*godo.Droplet, error) {
+	c.base.SendLog(fmt.Sprintf("Fetching droplets for %s", c.base.Name))
+	droplets := make([]*godo.Droplet, 0, c.base.NumInstances)
+	for _, id := range c.DropletIDs {
+		dr, _, err := c.client.Droplets.Get(int(id))
+		if err != nil {
+			return nil, err
+		}
+		droplets = append(droplets, dr.Droplet)
+	}
+	return droplets, nil
+}
+
+func (c *DigitalOceanCluster) configureDomain() error {
+	c.base.SendLog("Configuring domain")
+	instanceIP := c.base.InstanceIPs[0]
+	dr, _, err := c.client.Domains.Create(&godo.DomainCreateRequest{
+		Name:      c.base.Domain.Name,
+		IPAddress: instanceIP,
+	})
+	if err != nil {
+		return err
+	}
+	domainName := dr.Domain.Name
+	for i, ip := range c.base.InstanceIPs {
+		if i == 0 {
+			// An A record already exists via the create domain request
+			continue
+		}
+		_, _, err := c.client.Domains.CreateRecord(domainName, &godo.DomainRecordEditRequest{
+			Type: "A",
+			Name: fmt.Sprintf("%s.", domainName),
+			Data: ip,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	_, _, err = c.client.Domains.CreateRecord(domainName, &godo.DomainRecordEditRequest{
+		Type: "CNAME",
+		Name: fmt.Sprintf("*.%s.", domainName),
+		Data: fmt.Sprintf("%s.", domainName),
+	})
+	return err
+}
+
+func (c *DigitalOceanCluster) installFlynn() error {
+	c.base.SendLog("Installing flynn")
+
+	startScript, discoveryToken, err := c.base.genStartScript(c.base.NumInstances)
+	if err != nil {
+		return err
+	}
+	c.base.DiscoveryToken = discoveryToken
+	if err := c.base.saveField("DiscoveryToken", discoveryToken); err != nil {
+		return err
+	}
+	c.startScript = startScript
+
+	sshConfig, err := c.base.sshConfig()
+	if err != nil {
+		return err
+	}
+
+	instanceIPs := c.base.InstanceIPs
+	errChan := make(chan error)
```

Comment:
Add a buffer so that sends don't block and leak goroutines if there is an error.